Junos: autonomous-system support changes

rohitthakur2590 opened the pull request:
Signed-off-by: Rohit Thakur <rohitthakur2590@outlook.com>

##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
- routing-options autonomous-system implementation
- docs updated as per new states 
- integration tests updated w.r.t asn changes
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request
- Docs Pull Request
 

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
junos_bgp_global
##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```paste below

```

NilashishC commented:
Existing config:
```
"gathered": {
   "advertise_peer_as": true,
   "as_number": "65563"
},
```
```yaml
  tasks:
    - junipernetworks.junos.junos_bgp_global:
        config:
          as_number: "65563"
        state: replaced
```
This task results in traceback with rpc-error which I haven't been able to decipher. Expected behaviour would be `advertise_peer_as` removed (i.e., defaulted to False).

NilashishC commented:
```yaml
---
- hosts: junos
  gather_facts: no
  tasks:
    - junipernetworks.junos.junos_bgp_global:
        config:
          as_number: "65563"
        state: merged

    - junipernetworks.junos.junos_bgp_global:
        state: deleted
```

The second `deleted` task results in a traceback. Expected result would be to return green (i.e., no changes) since there are no parameters configured that global module handles except for `as_number`, which we will not negate since this is not purged.

NilashishC commented on plugins/modules/junos_bgp_global.py:
```diff
@@ -1133,7 +1139,18 @@
+      address-family related configurations under BGP context.
+    - Running states I(deleted) and I(replaced) will result in an error if there
+      are address-family configuration lines present under a neighbor.Please use the
+      M(junipernetworks.junos.junos_bgp_af) or M(junipernetworks.junos.junos_bgp_groups)
```
```suggestion
      M(junipernetworks.junos.junos_bgp_address_family) or M(junipernetworks.junos.junos_bgp_groups)
```

NilashishC replied:
We should also be throwing an error if the module attempts to delete a neighbor which has AF configurations like neighbor 192.168.1.10 and  neighbor 192.163.6.4 in this case:
```
ansible@an-vsrx-01# show protocols bgp | display set
set protocols bgp advertise-peer-as
set protocols bgp group external-peers neighbor 10.0.0.1 peer-as 65564
set protocols bgp group external-peers neighbor 10.0.0.2 peer-as 6556764
set protocols bgp group internal-peers type internal
set protocols bgp group internal-peers local-address 192.168.6.5
set protocols bgp group internal-peers cluster 192.168.6.5
set protocols bgp group internal-peers neighbor 192.163.6.4 family inet unicast
set protocols bgp group internal-peers neighbor 192.168.40.4
set protocols bgp group internal-peers neighbor 192.168.0.1
set protocols bgp group internal-peers neighbor 192.168.5.5
set protocols bgp group internal-peers neighbor 192.168.1.10 family inet unicast accepted-prefix-limit maximum 120
```

rohitthakur2590 replied:
this would be implemented in junos_bgp_group. not required in bgp_global.

NilashishC commented on plugins/modules/junos_bgp_global.py:
```diff
@@ -1404,36 +1433,42 @@
+# bmp {
+#     monitor enable;
 # }
+# add-path-display-ipv4-address;
```
An example that shows neighbor configuration and as_number would be great.

rohitthakur2590 replied:
yeah, that would be good to have that updated for merged and neighbor for delete operation example.

NilashishC commented on plugins/modules/junos_bgp_global.py:
```diff
@@ -1442,6 +1477,13 @@
+#
+#[edit]
+
 
```
In the example with parsed, the xml has as_number set https://github.com/ansible-collections/junipernetworks.junos/blob/c47e270ebd9d0e7eafc6c4cac0bad804c32305af/plugins/modules/junos_bgp_global.py#L1643, but the same is not shown in parsed data.

rohitthakur2590 replied:
yeah, I'll update the parsed opeartion

NilashishC commented:
There doesn't seem to be an option to configure neighbors. For example, we should be able to configure the following with this module.

```
ansible@an-vsrx-01# edit protocols bgp group external-peers
ansible@an-vsrx-01# set neighbor 10.0.0.1 peer-as 65564
ansible@an-vsrx-01# set neighbor 10.0.0.2 peer-as 65564
ansible@an-vsrx-01# show protocols bgp | display set
set protocols bgp advertise-peer-as
set protocols bgp group external-peers neighbor 10.0.0.1 peer-as 65564
set protocols bgp group external-peers neighbor 10.0.0.2 peer-as 6556764
```

NilashishC commented:
The global module should be able to configure at least basic top level neighbor settings such as the follows:
```
ansible@an-vsrx-01# set protocols bgp group internal-peers neighbor 192.163.6.4 ?
Possible completions:
  <[Enter]>            Execute this command
  accept-remote-nexthop  Allow import policy to specify a non-directly connected next-hop
  add-path-display-ipv4-address  Display add-path path-id in IPv4 address format
> advertise-bgp-static  Advertise bgp-static routes
> advertise-external   Advertise best external routes
  advertise-inactive   Advertise inactive routes
  advertise-peer-as    Advertise routes received from the same autonomous system
+ apply-groups         Groups from which to inherit configuration data
+ apply-groups-except  Don't inherit configuration data from these groups
  as-override          Replace neighbor AS number with our AS number
  authentication-algorithm  Authentication algorithm name
  authentication-key   MD5 authentication key
  authentication-key-chain  Key chain name
> bfd-liveness-detection  Bidirectional Forwarding Detection (BFD) options
> bgp-error-tolerance  Handle BGP malformed updates softly
> bmp                  Specific settings to override the routing-options settings
  cluster              Cluster identifier
  damping              Enable route flap damping
  description          Text description
> egress-te            Use Egress Peering traffic engineering
  enforce-first-as     Enforce first AS in AS-path is the neighbor's AS
+ export               Export policy
> family               Protocol family for NLRIs in updates
  forwarding-context   Routing-instance used for data-forwarding and transport-session
> graceful-restart     BGP graceful restart options
  hold-time            Hold time used when negotiating with a peer
> idle-after-switch-over  Stop peer session from coming up after nonstop-routing switch-over
+ import               Import policy
  include-mp-next-hop  Include NEXT-HOP attribute in multiprotocol updates
  ipsec-sa             IPSec SA name
  keep                 How to retain routes in the routing table
  local-address        Address of local end of BGP session
> local-as             Local autonomous system number
  local-interface      Local interface for IPv6 link local EBGP peering
  local-preference     Value of LOCAL_PREF path attribute
  log-updown           Log a message for peer state transitions
> metric-out           Route metric sent in MED
  mtu-discovery        Enable TCP path MTU discovery
> multihop             Configure an EBGP multihop session
> multipath            Allow load sharing among multiple BGP paths
  no-advertise-peer-as  Don't advertise routes received from the same autonomous system
  no-aggregator-id     Set router ID in aggregator path attribute to 0
  no-client-reflect    Disable intracluster route redistribution
  out-delay            How long before exporting routes from routing table (0..65535)
> outbound-route-filter  Dynamically negotiated cooperative route filtering
  passive              Do not send open messages to a peer
  peer-as               Autonomous system number in plain number or 'higher 16bits'.'Lower 16 bits' (asdot notation) format
  preference           Preference value
> remove-private       Remove well-known private AS numbers
  rfc6514-compliant-safi129  Compliant with RFC6514 SAFI129 format
  tcp-aggressive-transmission  Enable aggressive transmission of pure TCP ACKs and retransmissions
  tcp-mss              Maximum TCP segment size (1..4096)
> traceoptions         Trace options for BGP
  ttl                  TTL value for the single-hop peer
  unconfigured-peer-graceful-restart  BGP unconfigured peer graceful restart options
  vpn-apply-export     Apply BGP export policy when exporting VPN routes
```
As discussed neighbor_address_family module should be able to configure neighbor AF settings such as the follows:
```
ansible@an-vsrx-01# set protocols bgp group internal-peers neighbor 192.163.6.4 family ?
Possible completions:
> evpn                 EVPN NLRI parameters
> inet                 IPv4 NLRI parameters
> inet-mdt             IPv4 Multicast Distribution Tree (MDT) NLRI parameters
> inet-mvpn            IPv4 MVPN NLRI parameters
> inet-vpn             IPv4 Layer 3 VPN NLRI parameters
> inet6                IPv6 NLRI parameters
> inet6-mvpn           IPv6 MVPN NLRI parameters
> inet6-vpn            IPv6 Layer 3 VPN NLRI parameters
> iso-vpn              ISO Layer 3 VPN NLRI parameters
> l2vpn                MPLS-based Layer 2 VPN and VPLS NLRI parameters
> route-target         Route target NLRI used for VPN route filtering
> traffic-engineering  Traffic Engineering (BGP-TE) NLRI parameters
[edit]
ansible@an-vsrx-01# set protocols bgp group internal-peers neighbor 192.163.6.4 family inet unicast ?
Possible completions:
  <[Enter]>            Execute this command
> accepted-prefix-limit  Limit maximum number of prefixes accepted from a peer
> add-path             Advertise multiple paths to peer
> aigp                 Allow sending and receiving of AIGP attribute
+ apply-groups         Groups from which to inherit configuration data
+ apply-groups-except  Don't inherit configuration data from these groups
  damping              Enable route flap damping
> delay-route-advertisements  Delay route updates for this family until FIB-sync
  extended-nexthop     Extended nexthop encoding
> graceful-restart     BGP graceful restart options
  local-ipv4-address   Local IPv4 address
> loops                Allow local AS in received AS paths
  no-install           Dont install received routes in forwarding
> output-queue-priority  Default output-queue to assign updates to
> prefix-limit         Limit maximum number of prefixes from a peer
> rib-group            Routing table group
> route-refresh-priority  Default output-queue to assign route refreshes to
> topology             Multi topology routing tables
> withdraw-priority    Default output-queue to assign withdrawn routes to
```

rohitthakur2590 commented:
> Existing config:
> 
> ```
> "gathered": {
>    "advertise_peer_as": true,
>    "as_number": "65563"
> },
> ```
> 
> ```yaml
>   tasks:
>     - junipernetworks.junos.junos_bgp_global:
>         config:
>           as_number: "65563"
>         state: replaced
> ```
> 
> This task results in traceback with rpc-error which I haven't been able to decipher. Expected behaviour would be `advertise_peer_as` removed (i.e., defaulted to False).


1. Did you try with this PR ? , I am trying with this and it works fine
- So could you please share that rpc-error. we can analyze that to know why its giving error for you.
- So regarding expected behaviour , for abouve `initial config` with `replaced `  operation the `expected behaviour`
for junos should be :
```
  "after": {
        "as_number": "65563"
    },
    "before": {
        "advertise_peer_as": true,
        "as_number": "65563"
    },
```
so we gonna left with `routing-option` `as_number`. So we just don't have  `advertise_peer_as` configured anymore and no need to fetch that in facts as false.

rohitthakur2590 commented:
> ```yaml
> ---
> - hosts: junos
>   gather_facts: no
>   tasks:
>     - junipernetworks.junos.junos_bgp_global:
>         config:
>           as_number: "65563"
>         state: merged
> 
>     - junipernetworks.junos.junos_bgp_global:
>         state: deleted
> ```
> 
> The second `deleted` task results in a traceback. Expected result would be to return green (i.e., no changes) since there are no parameters configured that global module handles except for `as_number`, which we will not negate since this is not purged.
> 
> ```yaml
> ---
> - hosts: junos
>   gather_facts: no
>   tasks:
>     - junipernetworks.junos.junos_bgp_global:
>         config:
>           as_number: "65563"
>           advertise_peer_as: True
>         state: merged
> 
>     - junipernetworks.junos.junos_bgp_global:
>         state: deleted
> ```
> 
> The second `deleted` task results in a traceback. Expected result would be to negate `advertise_peer_as` only, keeping `as_number` intact.

so yeah as per the current behavior of BGP GLOBAL in Junos, if there are only global attributes and we use  `delete` operation 
, we can get rid of autonomous-system as well(the other option would be with purged which not only delete the autonomous-system config but Junos bgp resource as well ), but let's say if we have any neighbor configured with the association to autonomous-system, then we gonna get the error thrown by the Junos network device.

So looking at the behavior of Junos platform where you can have BGP global and address_family configures without even configuring the autonomous-system attributes, which is not the case for other platforms where all BGP config is within the context of as_number,  that makes sense for Junos.

So even if we just want to keep it aligned, one thing I can do here is I can remove the delete as_number handling from bgp_global which is the only module actually responsible for configuring as_number, and lets only purge handle that.
So if we see Juno's behavior Purged is not that much game changer here.

rohitthakur2590 commented:
> There doesn't seem to be an option to configure neighbors. For example, we should be able to configure the following with this module.
> 
> ```
> ansible@an-vsrx-01# edit protocols bgp group external-peers
> ansible@an-vsrx-01# set neighbor 10.0.0.1 peer-as 65564
> ansible@an-vsrx-01# set neighbor 10.0.0.2 peer-as 65564
> ansible@an-vsrx-01# show protocols bgp | display set
> set protocols bgp advertise-peer-as
> set protocols bgp group external-peers neighbor 10.0.0.1 peer-as 65564
> set protocols bgp group external-peers neighbor 10.0.0.2 peer-as 6556764
> ```
> 
> Options documented in this [guide](https://www.juniper.net/documentation/en_US/junos/topics/topic-map/bgp-rr.html#id-example-configuring-a-route-reflector) for setting up Route Reflectors also do not seem to be covered.
> 
> ```
> ansible@an-vsrx-01# show protocols bgp | display set
> set protocols bgp advertise-peer-as
> set protocols bgp group internal-peers type internal
> set protocols bgp group internal-peers local-address 192.168.6.5
> set protocols bgp group internal-peers cluster 192.168.6.5
> set protocols bgp group internal-peers neighbor 192.163.6.4
> set protocols bgp group internal-peers neighbor 192.168.40.4
> set protocols bgp group internal-peers neighbor 192.168.0.1
> set protocols bgp group internal-peers neighbor 192.168.5.5
> ```

So as per Juno's platform behaviour, neighbours are associated with BGP groups,
so for that, we have junos_bgp_groups resource module lines up, which will take care of all the associated neighbours and group leaf nodes.

and neighbor_address_family will take care of family attributes for each neighbour.

NilashishC commented:
> > There doesn't seem to be an option to configure neighbors. For example, we should be able to configure the following with this module.
> > ```
> > ansible@an-vsrx-01# edit protocols bgp group external-peers
> > ansible@an-vsrx-01# set neighbor 10.0.0.1 peer-as 65564
> > ansible@an-vsrx-01# set neighbor 10.0.0.2 peer-as 65564
> > ansible@an-vsrx-01# show protocols bgp | display set
> > set protocols bgp advertise-peer-as
> > set protocols bgp group external-peers neighbor 10.0.0.1 peer-as 65564
> > set protocols bgp group external-peers neighbor 10.0.0.2 peer-as 6556764
> > ```
> > 
> > 
> > Options documented in this [guide](https://www.juniper.net/documentation/en_US/junos/topics/topic-map/bgp-rr.html#id-example-configuring-a-route-reflector) for setting up Route Reflectors also do not seem to be covered.
> > ```
> > ansible@an-vsrx-01# show protocols bgp | display set
> > set protocols bgp advertise-peer-as
> > set protocols bgp group internal-peers type internal
> > set protocols bgp group internal-peers local-address 192.168.6.5
> > set protocols bgp group internal-peers cluster 192.168.6.5
> > set protocols bgp group internal-peers neighbor 192.163.6.4
> > set protocols bgp group internal-peers neighbor 192.168.40.4
> > set protocols bgp group internal-peers neighbor 192.168.0.1
> > set protocols bgp group internal-peers neighbor 192.168.5.5
> > ```
> 
> So as per Juno's platform behaviour, neighbours are associated with BGP groups,
> so for that, we have junos_bgp_groups resource module lines up, which will take care of all the associated neighbours and group leaf nodes.
> 
> and neighbor_address_family will take care of family attributes for each neighbour.

Is there a reason why we need a separate module for BGP neighbors and not handle it in the global module? I'm just trying to make sure we keep the modules as consistent as possible across all the platforms.

NilashishC commented:
> > Existing config:
> > ```
> > "gathered": {
> >    "advertise_peer_as": true,
> >    "as_number": "65563"
> > },
> > ```
> > 
> > 
> > ```yaml
> >   tasks:
> >     - junipernetworks.junos.junos_bgp_global:
> >         config:
> >           as_number: "65563"
> >         state: replaced
> > ```
> > 
> > 
> > This task results in traceback with rpc-error which I haven't been able to decipher. Expected behaviour would be `advertise_peer_as` removed (i.e., defaulted to False).
> 
> 1. Did you try with this PR ? , I am trying with this and it works fine
> 
> * So could you please share that rpc-error. we can analyze that to know why its giving error for you.
> * So regarding expected behaviour , for abouve `initial config` with `replaced `  operation the `expected behaviour`
>   for junos should be :
> 
> ```
>   "after": {
>         "as_number": "65563"
>     },
>     "before": {
>         "advertise_peer_as": true,
>         "as_number": "65563"
>     },
> ```
> 
> so we gonna left with `routing-option` `as_number`. So we just don't have `advertise_peer_as` configured anymore and no need to fetch that in facts as false.

I agree with the expected behaviour, that's what I meant too. Here's a gist of the traceback I'm seeing. It also has other details like JUNOS version and existing state - https://gist.github.com/NilashishC/3e3a5f9133c6d1403e034e95238a7d07

rohitthakur2590 commented:
> > > There doesn't seem to be an option to configure neighbors. For example, we should be able to configure the following with this module.
> > > ```
> > > ansible@an-vsrx-01# edit protocols bgp group external-peers
> > > ansible@an-vsrx-01# set neighbor 10.0.0.1 peer-as 65564
> > > ansible@an-vsrx-01# set neighbor 10.0.0.2 peer-as 65564
> > > ansible@an-vsrx-01# show protocols bgp | display set
> > > set protocols bgp advertise-peer-as
> > > set protocols bgp group external-peers neighbor 10.0.0.1 peer-as 65564
> > > set protocols bgp group external-peers neighbor 10.0.0.2 peer-as 6556764
> > > ```
> > > 
> > > 
> > > Options documented in this [guide](https://www.juniper.net/documentation/en_US/junos/topics/topic-map/bgp-rr.html#id-example-configuring-a-route-reflector) for setting up Route Reflectors also do not seem to be covered.
> > > ```
> > > ansible@an-vsrx-01# show protocols bgp | display set
> > > set protocols bgp advertise-peer-as
> > > set protocols bgp group internal-peers type internal
> > > set protocols bgp group internal-peers local-address 192.168.6.5
> > > set protocols bgp group internal-peers cluster 192.168.6.5
> > > set protocols bgp group internal-peers neighbor 192.163.6.4
> > > set protocols bgp group internal-peers neighbor 192.168.40.4
> > > set protocols bgp group internal-peers neighbor 192.168.0.1
> > > set protocols bgp group internal-peers neighbor 192.168.5.5
> > > ```
> > 
> > 
> > So as per Juno's platform behavior, neighbors are associated with BGP groups,
> > so for that, we have junos_bgp_groups resource module lines up, which will take care of all the associated neighbors and group leaf nodes.
> > and neighbor_address_family will take care of family attributes for each neighbor.
> 
> Is there a reason why we need a separate module for BGP neighbors and not handle it in the global module? I'm just trying to make sure we keep the modules as consistent as possible across all the platforms.

yes because it would make bgp_global_module more complex to have the BGP group lists with all the nested node structure there, so for Junos it makes resource module use easy for the user,  and on other hand reduce the complexity for bgp.* resource module with a separate `junos_bgp_groups` module.
I agree with you when it comes to consistency across all the platforms, and I think Junos is pretty aligned with other modules w.r.t category/structure of other platform bgp modules, but then some things need to be taken care with specific platform behaviour as well.

rohitthakur2590 commented:
> > > Existing config:
> > > ```
> > > "gathered": {
> > >    "advertise_peer_as": true,
> > >    "as_number": "65563"
> > > },
> > > ```
> > > 
> > > 
> > > ```yaml
> > >   tasks:
> > >     - junipernetworks.junos.junos_bgp_global:
> > >         config:
> > >           as_number: "65563"
> > >         state: replaced
> > > ```
> > > 
> > > 
> > > This task results in traceback with rpc-error which I haven't been able to decipher. Expected behaviour would be `advertise_peer_as` removed (i.e., defaulted to False).
> > 
> > 
> > 
> > 1. Did you try with this PR ? , I am trying with this and it works fine
> > 
> > 
> > * So could you please share that rpc-error. we can analyze that to know why its giving error for you.
> > * So regarding expected behaviour , for abouve `initial config` with `replaced `  operation the `expected behaviour`
> >   for junos should be :
> > 
> > ```
> >   "after": {
> >         "as_number": "65563"
> >     },
> >     "before": {
> >         "advertise_peer_as": true,
> >         "as_number": "65563"
> >     },
> > ```
> > 
> > 
> > so we gonna left with `routing-option` `as_number`. So we just don't have `advertise_peer_as` configured anymore and no need to fetch that in facts as false.
> 
> I agree with the expected behaviour, that's what I meant too. Here's a gist of the traceback I'm seeing. It also has other details like JUNOS version and existing state - https://gist.github.com/NilashishC/3e3a5f9133c6d1403e034e95238a7d07

so we have been testing this for version 18.4R1;
That's why you are getting a different result at your end.

NilashishC commented:
> > > > Existing config:
> > > > ```
> > > > "gathered": {
> > > >    "advertise_peer_as": true,
> > > >    "as_number": "65563"
> > > > },
> > > > ```
> > > > 
> > > > 
> > > > ```yaml
> > > >   tasks:
> > > >     - junipernetworks.junos.junos_bgp_global:
> > > >         config:
> > > >           as_number: "65563"
> > > >         state: replaced
> > > > ```
> > > > 
> > > > 
> > > > This task results in traceback with rpc-error which I haven't been able to decipher. Expected behaviour would be `advertise_peer_as` removed (i.e., defaulted to False).
> > > 
> > > 
> > > 
> > > 1. Did you try with this PR ? , I am trying with this and it works fine
> > > 
> > > 
> > > * So could you please share that rpc-error. we can analyze that to know why its giving error for you.
> > > * So regarding expected behaviour , for abouve `initial config` with `replaced `  operation the `expected behaviour`
> > >   for junos should be :
> > > 
> > > ```
> > >   "after": {
> > >         "as_number": "65563"
> > >     },
> > >     "before": {
> > >         "advertise_peer_as": true,
> > >         "as_number": "65563"
> > >     },
> > > ```
> > > 
> > > 
> > > so we gonna left with `routing-option` `as_number`. So we just don't have `advertise_peer_as` configured anymore and no need to fetch that in facts as false.
> > 
> > 
> > I agree with the expected behaviour, that's what I meant too. Here's a gist of the traceback I'm seeing. It also has other details like JUNOS version and existing state - https://gist.github.com/NilashishC/3e3a5f9133c6d1403e034e95238a7d07
> 
> so we have been testing this for version 18.4R1;
> That's why you are getting a different result at your end.

Okay, in that case, I believe we can have a separate discussing on network OS version support. Rest of the changes looks good to me. Thank you.
